test/distributed/test_c10d_nccl.py

```diff
@@ -410,6 +410,27 @@ def test_alltoall_ops_with_cudafree_race(self):
             work.wait()
         torch.cuda.synchronize(local_device)
 
+    @requires_nccl()
+    @skip_but_pass_in_sandcastle_if(torch.cuda.device_count() < 2, "NCCL test requires 2+ GPUs")
+    def test_allreduce_in_cudagraph(self):
+        store = c10d.FileStore(self.file_name, self.world_size)
+        pg = self._create_process_group_nccl(store, self.opts())
+        local_device_idx = self.rank_to_GPU[self.rank][0]
+        with torch.cuda.device(local_device_idx):
```

Comment:
I think we should figure out why a device context is needed here. If it is specifically needed for collectives, we should probably investigate more to see if there are other bugs in `ProcessGroupNCCL`. Can be in follow-up PRs.

Reply:
Sorry for the delay in replying to this comment. This context manager is very much needed, and my confusion about this is what was broken in https://github.com/pytorch/pytorch/pull/102542 . This is not evidence of a bug. The call to `torch.cuda.graph` by default records actions on the current stream of the current device. When the tests run without this context manager selecting the default device, at least one of the processes typically won't have the default device being `local_device_index`, which means nothing will get recorded in that process. 

(There are actually two types of failures which can happen. If neither process records, then the allreduce in the capture block will actually happen, so the second assert will fail as 4 != 2. If one process records, then one process will try to allreduce when the other one doesn't, resulting in a hang.)